addons/account/models/partner.py

```diff
@@ -434,6 +434,26 @@ def _compute_bank_count(self):
         for partner in self:
             partner.bank_account_count = mapped_data.get(partner.id, 0)
 
+    @api.constrains('company_id')
+    def _check_company_and_amls(self):
+        query = """SELECT aml.id
+                FROM account_account as account,
+                    account_move as move,
+                    account_move_line as aml
+                WHERE
+                    aml.move_id = move.id
+                    AND aml.account_id = account.id
+                    AND aml.partner_id in %s
+                    AND (aml.reconciled IS NULL or aml.reconciled = false)
+                    AND (account.deprecated IS NULL or account.deprecated = false)
+                    AND move.state = 'posted'
+                    AND EXISTS (SELECT 1
+                                FROM res_partner as partner
+                                WHERE partner.id = aml.partner_id AND partner.company_id <> aml.company_id)"""
+        self.env.cr.execute(query, (tuple(self.ids),))
+        if self.env.cr.fetchall():
+            raise ValidationError(_("You can't change the company of a partner if the latter still has some unreconciled account moves."))
```

Comment:
This needs to be done only for records where `company_id` is set. You can always remove the value if it was set before.

Reply:
Also, I'm not sure this is only related to reconciliation either.
For instance:
* Have two companies A and B
* Have a partner P without company set
* Create an invoice for P in A
* Set company B on P
* Try to open the invoice in A

--> Result: access restriction

This is not even only related to accounting but to almost all documents everywhere in Odoo.

Reply:
So the idea would be rather to prevent any company change if the customer already has some invoices ?

Reply:
Yes, otherwise the other companies won't be able to open their invoices anymore...